# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/ipc/CanvasTranslator.cpp
# Commit: 57fc38773fe3
# Full Hash: 57fc38773fe37f21748092e3e58fdf5265e85c0d
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2023-08-30 04:26:43
# Regressor Bug: 1841891
# File Overlap Count: 3
# Description:
#   Bug 1841891: Only use a larger canvas ring buffer when in the foreground. r=jrmuizel
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D186544
# ==============================================================================

diff -r 082442e3b925 -r 57fc38773fe3 gfx/layers/ipc/CanvasTranslator.cpp
--- a/gfx/layers/ipc/CanvasTranslator.cpp	Tue Aug 29 19:15:34 2023 +0000
+++ b/gfx/layers/ipc/CanvasTranslator.cpp	Tue Aug 29 19:17:45 2023 +0000
@@ -157,6 +157,19 @@
   return RecvResumeTranslation();
 }
 
+ipc::IPCResult CanvasTranslator::RecvNewBuffer(
+    ipc::SharedMemoryBasic::Handle&& aReadHandle) {
+  // We need to set the new buffer on the transaltion queue to be sure that the
+  // drop buffer event has been processed.
+  MOZ_ALWAYS_SUCCEEDS(mTranslationTaskQueue->Dispatch(NS_NewRunnableFunction(
+      "CanvasTranslator SetNewBuffer",
+      [self = RefPtr(this), readHandle = std::move(aReadHandle)]() mutable {
+        self->mStream->SetNewBuffer(std::move(readHandle));
+      })));
+
+  return RecvResumeTranslation();
+}
+
 ipc::IPCResult CanvasTranslator::RecvResumeTranslation() {
   if (mDeactivated) {
     // The other side might have sent a resume message before we deactivated.
@@ -171,6 +184,9 @@
 }
 
 void CanvasTranslator::StartTranslation() {
+  MOZ_RELEASE_ASSERT(mStream->IsValid(),
+                     "StartTranslation called before buffer has been set.");
+
   if (!TranslateRecording() && GetIPCChannel()->CanSend()) {
     MOZ_ALWAYS_SUCCEEDS(mTranslationTaskQueue->Dispatch(
         NewRunnableMethod("CanvasTranslator::StartTranslation", this,
@@ -247,7 +263,7 @@
   MOZ_ASSERT(CanvasThreadHolder::IsInCanvasWorker());
 
   uint8_t eventType = mStream->ReadNextEvent();
-  while (mStream->good()) {
+  while (mStream->good() && eventType != kDropBufferEventType) {
     bool success = RecordedEvent::DoWithEventFromStream(
         *mStream, static_cast<RecordedEvent::EventType>(eventType),
         [&](RecordedEvent* recordedEvent) -> bool {